### PR TITLE
Move src/test to the root

### PR DESCRIPTION
**Pull request by albertlarsan68**
See MCP at rust-lang/compiler-team#573

There may be more changes needed.

The first commit is just the move of the files:
You can check that the first commit did not do anything else than renames by running 
```
git diff --diff-filter=r -M100% <rust-lang remote>/master <first commit hash>
```
The output should be empty, because the filter excludes renames, and the match threshold for qualifying a rename is 100%.

The second one is mostly a "find and replace" of `src/test` to `tests` and whatever is needed to make CI pass.

What is left to do:
---

- [x] Move directory
- [x] Change references to `src/test`
    - [x] Change references in-tree
    - [x] Change references in submodules / out-of-tree docs
- [x] Make CI pass:
    - [x] Fix tidy
    - [x] Fix tests
    - [x] Bless tests if needed (shouldn't normally)
- [x] Merge it !

### CODE REVIEW

**Comment by jyn514** [comment]
@albertlarsan68 in addition to fixing the CI failures, please also make sure you update `src/bootstrap/test.rs` to add `.path("test/suite_name")` to all the test suites. I think for now we should *keep* `x test src/test/ui` as an alias for `x test test/ui`,  but we should definitely support the latter.

**Comment by petrochenkov** [comment]
I guess the only question is whether to keep the name `test`, or rename it to `tests` which is the cargo convention for integration tests.

**Comment by jyn514** [comment]
I would prefer `tests`, both for consistency with cargo and to avoid ambiguity with `library/test` in case we ever want to make `x test tests` work (i.e. run only compiletest suites and not unit tests or tidy).

**Comment by the8472** [comment]
This should split into multiple commits? one that only moves files and the rest for adjustments to the tools. That'll make it possible to review the code changes on GH.

**Comment by albertlarsan68** [comment]
You can check that the first commit did not do anything else than renames by running 
```
git diff --diff-filter=r -M100% <rust-lang remote>/master <first commit hash>
```

**Comment by jyn514** [comment]
@albertlarsan68 you may want to temporarily disable the x86_64-gnu-tools job until #106449 lands to avoid CI spuriously failing while you're trying to fix other things.

Also, given the rebase problem, you may want to run the tests locally first instead of primarily using CI. If that's prohibitively slow I can look into getting you access to one of the cloud dev machines.

File: tests/mir-opt/building/custom/arbitrary_let.arbitrary_let.built.after.mir
Comment (jyn514) [inline tests/mir-opt/building/custom/arbitrary_let.arbitrary_let.built.after.mir:5]: This doesn't look right.

Comment (albertlarsan68) [reply tests/mir-opt/building/custom/arbitrary_let.arbitrary_let.built.after.mir:5]: Which commit is it ?

**Comment by albertlarsan68** [comment]
I can't reproduce the CI failure locally, it seems like the normalization code is broken only on CI, for some reason.

**Comment by albertlarsan68** [comment]
@jyn514 PR CI passed, what should I do next ?

**Comment by jyn514** [comment]
@albertlarsan68 no, it's not running tests, only building the compiler - I wanted to test if PGO was still working.

I think the next step is for me to do an initial review pass :)

**Comment by albertlarsan68** [comment]
Ping me when you're ready for me to reabse

**Comment by albertlarsan68** [comment]
I maintain a rebased branch at #106631, so that the constant force-pushes do not slow down your review work.

File: src/tools/tidy/src/edition.rs
Comment (jyn514) [inline src/tools/tidy/src/edition.rs:16]: this `COMPILER_TESTS.md` change is new, right? why did that change?

If you're trying to filter *only* the top-level directory, maybe you can do that directly?
```suggestion
                || path == "tests"
```

File: src/tools/tidy/src/error_codes_check.rs
Comment (jyn514) [inline src/tools/tidy/src/error_codes_check.rs:25]: this looks *super* sketchy, but not something you need to fix in this PR

File: src/tools/tidy/src/features.rs
Comment (jyn514) [inline src/tools/tidy/src/features.rs:87]: maybe just pass in `root_path` to avoid blowing up the number of arguments? but this PR is already quite large, I don't mind making that a follow-up.

**Review by jyn514 — APPROVED** [review]
this looks really good! I have a couple small nits which don't need to be fixed in this PR.

The only question I have is on how to merge this. nearly any PR that merges first will cause conflicts, so if you need to wait for my approval each time after rebasing it will take many tries (even with a very high priority).

how about this: can you leave a response to all my review comments, and if I'm happy with them (and the PR already has merge conflicts, which I expect to be the case) I'll delegate the approval to you. Then you can use `@bors r=jyn514` to approve *immediately* after you rebase without having to wait for me to have time.

**Comment by jyn514** [comment]
oh, please also squash all commits *other* than the first `src/test` -> `tests` rename into a single commit, i.e. there should be exactly 2 commits after you finish rebasing.

**Comment by calebcartwright** [comment]
Somewhat tangential aside related to https://github.com/rust-lang/rust/pull/106458#issuecomment-1376513443 but more broadly, it would be nice if we could find a way to minimize the number of unrelated changes to tools. I feel like I'm seeing a bit of an uptick of people making in-tree find+replace changes that are causing minor, but unnecessary churn (e.g. the diff to the rustfmt files)

To be clear I'm not requesting any changes to this PR, particularly given how prone it is to bit rot. Just wondering aloud if there's some additional text we can add to e.g. the dev guide to give more guidance about when changes are, or are not, required on subtrees.

**Comment by jyn514** [comment]
@calebcartwright I'm not sure why you see that as unnecessary churn? I guess we could leave the comments be, but it seems worse to me to have them point to the wrong place.

**Comment by calebcartwright** [comment]
> @calebcartwright I'm not sure why you see that as unnecessary churn? I guess we could leave the comments be, but it seems worse to me to have them point to the wrong place.

Happy to discuss in more detail offline if there's interest, but in short the minor change to the rustfmt source (e.g. https://github.com/rust-lang/rust/pull/106458/files#diff-7e242a2587d660745691ca2d767b3acc58ac163c80f5405fb097d72088d94303) is not necessary in order for the tests to continue passing, nor do I think there's any comments in/around that portion of the diff that pose a risk of confusion.

In a longer take: It's not an explicit reference to the rustc codebase and was presumably only updated due to a coincidental string match. The tests would continue passing regardless of whether the string was left as-is, with the update, changed to `"foo/bar"`, etc. 

Again it's minor, but every change will show up as an extra commit in the tool repo, is another commit we have to process when diffing versions/building out release notes, increases the chance of extra work during subtree syncs/merge conflicts, etc.

I mentioned it here because this was the most recent instance, but it's the type of thing we're starting to see at an increasing rate so I wanted to make a somewhat more meta-point about the pattern (not this nor any one particular instance). It seemed worthwhile to me to consider whether it's something worth addressing/attempting to avoid in the future, especially if it's happening with other tools too.

File: src/tools/tidy/src/edition.rs
Comment (albertlarsan68) [reply src/tools/tidy/src/edition.rs:16]: This doesn't work, because the path will never start with `tests`. It will at least start with a `/` or `./`. I do not know how to detect this any better

File: src/tools/tidy/src/features.rs
Comment (albertlarsan68) [reply src/tools/tidy/src/features.rs:87]: It would make another thing dependant on the place of the tests. It was already a big hunt to find all of those now, adding even more does not seem like a very good idea.

File: src/tools/tidy/src/error_codes_check.rs
Comment (albertlarsan68) [reply src/tools/tidy/src/error_codes_check.rs:25]: Then let's do this in another PR.

**Comment by albertlarsan68** [comment]
If you want to take a last look at the changes, there there are. If you want me to revert the changes to rustdoc, I can do them.

Concerning the "How to merge this" question, I think the best way would be to treeclose, and r+/rebase until this PR passes.

File: src/tools/tidy/src/edition.rs
Comment (albertlarsan68) [reply src/tools/tidy/src/edition.rs:16]: While investigating this issue, I just saw that I forgot to also include the tests directory for this check, thanks!

Here is an example of path this function gets: `/home/gh-albertlarsan68/rust/compiler/rustc_arena/src/tests.rs`, and they are all like this, so testing for exact match with `tests` does definitely not work.

**Comment by jyn514** [comment]
@bors r+ p=50 rollup=never delegate=albertlarsan68 

Don't expect this to pass the first time, but worth running past CI.

**Comment by jyn514** [comment]
@bors p=49

https://github.com/rust-lang/rust/pull/106667#issuecomment-1377389659

**Comment by jyn514** [comment]
Looks like the wasm tests on test-various are failing.

```
(node:internal/modules/run_main:76:12)
    at node:internal/main/run_main_module:17:47 {
  code: 'MODULE_NOT_FOUND',
  requireStack: []
}
------------------------------------------


---- [ui] checkout/tests/ui/union/union-trait-impl.rs#mirunsafeck stdout ----

error in revision `mirunsafeck`: test run failed!
status: exit status: 1
command: cd "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/union/union-trait-impl.mirunsafeck" && RUST_TEST_THREADS="16" "/node-v15.14.0-linux-x64/bin/node" "/src/etc/wasm32-shim.js" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/union/union-trait-impl.mirunsafeck/a.wasm"
stdout: none
--- stderr -------------------------------
node:internal/modules/cjs/loader:927
  throw err;
  ^

Error: Cannot find module '/src/etc/wasm32-shim.js'
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:924:15)
    at Function.Module._load (node:internal/modules/cjs/loader:769:27)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:76:12)
    at node:internal/main/run_main_module:17:47 {
  code: 'MODULE_NOT_FOUND',
  requireStack: []
}
```

**Comment by JohnTitor** [comment]
The diff is huge and I cannot check on web UI, but is there a tidy check that prevents contributors from adding a test to `src/test`?
A PR modifying an existing test now has a merge conflict e.g. https://github.com/rust-lang/rust/pull/106167 but a PR adding a new test doesn't e.g. https://github.com/rust-lang/rust/pull/106660, and I'm worried if a new test is added to `src/test` by accident.

**Comment by albertlarsan68** [comment]
No, there isn't.
Can you open an issue for it ?
I will add it.

**Comment by albertlarsan68** [comment]
Can you also open one to update the references in `triagebot.toml` if it isn't already done ?

**Comment by JohnTitor** [comment]
Sure, opened https://github.com/rust-lang/rust/issues/106724.

**Comment by JohnTitor** [comment]
> Can you also open one to update the references in `triagebot.toml` if it isn't already done ?

Couldn't find any reference:
image

**Comment by jhpratt** [comment]
@albertlarsan68 Can you double check tidy? [This error](https://github.com/rust-lang/rust/actions/runs/3895584890/jobs/6651085249) is definitely not accurate, as there are test files with the required lines.

**Comment by jyn514** [comment]
@jhpratt that error looks correct, the tests should be in `tests/ui/restrictions`, not `tests/restrictions`

**Comment by jhpratt** [comment]
Duh...thanks for the response!

**Pull request merged**
